### DISCUSSION TITLE
inlining frames: how's supposed to work?

### DISCUSSION THREAD

**Discussion by franz1981** [async-profiler/async-profiler#564] [Q&A] [answered]
![image](https://user-images.githubusercontent.com/13125299/158187666-7876ea92-f780-476c-94a5-94c66ba305d9.png)

This flamegraph shows that echo isn't inlined: its ancestor `echoNoTimeout` is marked with `@CompilerControl(CompilerControl.Mode.DONT_INLINE)` and compilers logs report that's 'echo` is correctly inlined; have I misunderstood what the inlining frames feature works (ie maybe reporting just top inlined frames, not mid-ones)?

I can create a minimal different benchmark/example if it helps

**Comment by apangin** [comment, marked as answer]
The current implementation highlights only top frames.
Async-profiler knows the topmost real Java frame and its type: interpreted or compiled. If it is compiled, all virtual frames above it are marked as inlined.
On your picture, `echo` and `echoNoTimeout` are in the middle of the stack, and shown in default color.

**Reply by apangin** [reply]
I'm investigating the possibility to color the entire flame graph. This will likely require a major change of the profiler architecture.

**Reply by franz1981** [reply]
Many thanks, so yep: I've misunderstood what the feature was delivering (that's an improvement already over the original frame type detection/visualization), but probably users should know about it or they'll believe mid-frames are not getting inlined
